docs/common_functions.md

```diff
@@ -0,0 +1,252 @@
+
+## include
+
+If you want to import some functionality that you use frequently without having to copy/paste it between files, you can use `include`:
+
+```lisp
+;; condition_codes.clvm
+(
+  (defconstant AGG_SIG_ME 50)
+  (defconstant CREATE_COIN 51)
+)
+```
+
+```lisp
+;;main.clvm
+(mod create_sig_coin (pubkey msg puzzle_hash amount)
+
+  (include "condition_codes.clvm")
+
+  (list (list AGG_SIG_ME pubkey msg) (list CREATE_COIN puzzle_hash amount))
+
+)
+```
+
+When running main.clvm with `run`, make sure to use the `-i` option to specify in which directories to look for files.
+
+Also note that the include files are a special format. All functions and constants defined go in a set of parentheses.
+
+## sha256tree1
+
+When puzzles are hashed, they are not simply serialized and passed to sha256.  Instead, we take the *tree hash* of the puzzle.
+
+Recall that every clvm program can be represented as a binary tree.  Every object is either an atom (a leaf of the tree) or a cons box (a branch of the tree).  When we hash the puzzle we start at the leaves of the tree and hash our way up, concatenating either a 1 or a 2 to denote that it's either an atom or a cons box.  Once a cons box is hashed, it becomes a new leaf to be hashed into its parent cons box and the process recurses.  Here's what that looks like in Chialisp:
+
+```lisp
+(defun sha256tree1
+   (TREE)
+   (if (l TREE)
+       (sha256 2 (sha256tree1 (f TREE)) (sha256tree1 (r TREE)))
+       (sha256 1 TREE)
+   )
+)
+```
+
+It is extremely useful to calculate tree hashes within a Chialisp puzzle.  You can assert puzzles of other coins, condense puzzles for easier signing, and make CREATE_COIN conditions that are dependent on some passed in data.  
+
+## Currying
+
+Currying is an extremely important concept in Chialisp that is responsible for almost the entirety of how state is stored in coins.  The idea is to pass in arguments to a puzzle *before* it is hashed.  When you curry, you commit to solution values so that the individual solving the puzzle cannot change them.  Let's take a look at how this is implemented in Chialisp:
+
+```lisp
+;; utility function used by curry
+(defun fix_curry_args (items core)
+ (if items
+     (qq (c (q . (unquote (f items))) (unquote (fix_curry_args (r items) core))))
+     core
+ )
+)
+
+; (curry sum (list 50 60)) => returns a function that is like (sum 50 60 ...)
+(defun curry (func list_of_args) (qq (a (q . (unquote func)) (unquote (fix_curry_args list_of_args (q . 1))))))
+```
+
+The reason this is so useful is because you want to be able to reuse puzzles, but not necessarily always have the same values used inside the puzzle.  You can't rely on the puzzle solver to honestly and correctly pass in the information you want to use, so you need to make sure it is passed in before they ever get the chance to solve it.
```

Comment:
Maybe expand / rephrase this?
I like to talk about this in terms of allowing us to create blueprints for puzzles.

Also maybe explain how it works technically, wrapping in an `(a)` with changed params so we can have that knowledge set up ready for uncurrying.